L1Trigger/TrackerTFP/src/DataFormats.cc

```diff
@@ -0,0 +1,817 @@
+    get<6>(data_) = format(Variable::cot, Process::sf).floating(stub.cot());
+    dataFormats_->convertTrack(data_, frame_.second, p_);
+  }
+
+  // construct TrackKFin from TTTrackRef
+  TrackKFin::TrackKFin(const TTTrackRef& ttTrackRef, const DataFormats* dataFormats, const TTBV& maybePattern, double phiT, double inv2R, double zT, double cot, int sectorPhi, int sectorEta) :
+    Track(ttTrackRef, dataFormats, Process::kfin, maybePattern, sectorPhi, sectorEta, phiT, inv2R, zT, cot),
+    stubs_(setup()->numLayers()),
+    hitPattern_(0, setup()->numLayers())
+  {
+    dataFormats_->convertTrack(data_, frame_.second, p_);
+  }
+
+  vector<TTStubRef> TrackKFin::ttStubRefs(const TTBV& hitPattern, const vector<int>& layerMap) const {
+    vector<TTStubRef> stubs;
+    stubs.reserve(hitPattern.count());
+    for (int layer = 0; layer < setup()->numLayers(); layer++)
+      if (hitPattern[layer])
+        stubs.push_back(stubs_[layer][layerMap[layer]]->ttStubRef());
+    return stubs;
+  }
+
+  // construct TrackKF from Frame
+  TrackKF::TrackKF(const FrameTrack& frame, const DataFormats* dataFormats) :
+    Track(frame, dataFormats, Process::kf) {}
+
+  // construct TrackKF from TrackKfin
+  TrackKF::TrackKF(const TrackKFin& track, double phiT, double inv2R, double zT, double cot) :
+    Track(track, false, track.sectorPhi(), track.sectorEta(), track.phiT(), track.inv2R(), track.cot(), track.zT())
+  {
+    get<0>(data_) = abs(inv2R) < dataFormats_->format(Variable::inv2R, Process::sf).base() / 2. && abs(phiT) < dataFormats_->format(Variable::phiT, Process::sf).base() / 2.;
+    get<3>(data_) += phiT;
+    get<4>(data_) += inv2R;
+    get<5>(data_) += cot;
+    get<6>(data_) += zT;
+    dataFormats_->convertTrack(data_, frame_.second, p_);
+  }
+
+  // conversion to TTTrack with given stubs
+  TTTrack<Ref_Phase2TrackerDigi_> TrackKF::ttTrack(const vector<StubKF>& stubs) const {
+    const double invR = -this->inv2R() * 2.;
+    const double phi0 = deltaPhi(this->phiT() - this->inv2R() * dataFormats_->chosenRofPhi() + setup()->baseSector() * (this->sectorPhi() - .5));
+    const double cot = this->cot() + setup()->sectorCot(this->sectorEta());
+    const double z0 = this->zT() - this->cot() * setup()->chosenRofZ();
+    TTBV hitVector(0, setup()->numLayers());
+    double chi2phi(0.);
+    double chi2z(0.);
+    vector<TTStubRef> ttStubRefs;
+    const int nLayer = stubs.size();
+    ttStubRefs.reserve(nLayer);
+    for (const StubKF& stub : stubs) {
+      hitVector.set(stub.layer());
+      const TTStubRef& ttStubRef = stub.ttStubRef();
+      chi2phi += pow(stub.phi(), 2) / setup()->v0(ttStubRef, this->inv2R());
+      chi2z += pow(stub.z(), 2) / setup()->v1(ttStubRef, cot);
+      ttStubRefs.push_back(ttStubRef);
+    }
+    static constexpr int nParPhi = 2;
+    static constexpr int nParZ = 2;
+    static constexpr int nPar = nParPhi + nParZ;
+    const int dofPhi = nLayer - nParPhi;
+    const int dofZ = nLayer - nParZ;
+    chi2phi /= dofPhi;
```

Comment:
Is this a bug? TTTrack expects to be given chi2rphi and chi2rz, but not normalized to degrees of freedom.

Reply:
I thought the f/w is outputting chi2s odf, is that wrong?

Reply:
The constructor of TTTrack expects chi2rphi & chi2rz, not normalized to dof. It is true that the L1 tracking will now output chi2/dof, but that doesn't affect the call to the TTTrack constructor.